File: problem-solving/_lc-300-medium-longest-increasing-subsequence/1-top-down.py
"""
https://leetcode.com/problems/longest-increasing-subsequence/

Complexity
----------
Time: O(n^2)
Space: O(n^2)

NOTE: This solution is not accepted, gives "Memory limit exceeded"
"""

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        cache = {}
        def _lengthOfLIS(prev, idx):
            if idx == len(nums):
                return 0
            if (prev, idx) in cache:
                return cache[(prev, idx)]
            taken = 0
            notTaken = _lengthOfLIS(prev, idx + 1)
            if prev < nums[idx]:
                taken = 1 + _lengthOfLIS(nums[idx], idx + 1)
            cache[(prev, idx)] = max(taken, notTaken)
            return cache[(prev, idx)]

        return _lengthOfLIS(-math.inf, 0)

File: problem-solving/_lc-300-medium-longest-increasing-subsequence/1-top-down.go
/*
https://leetcode.com/problems/longest-increasing-subsequence/

Complexity
----------
Time: O(n^2)
Space: O(n^2)

NOTE: This solution is not accepted, gives "Time limit exceeded"
*/

func lengthOfLIS(nums []int) int {
    cache := make(map[string]int)
    var F func(int, int) int
    F = func (prev, idx int) int {
        if idx == len(nums) {
            return 0
        }
        key := fmt.Sprintf("%d-%d", prev, idx)
        if v, ok := cache[key]; ok {
            return v
        }
        notTaken := F(prev, idx + 1)
        taken := 0
        if nums[idx] > prev {
            taken = 1 + F(nums[idx], idx + 1)
        }
        if taken > notTaken {
            cache[key] = taken
        } else {
            cache[key] = notTaken
        }
        return cache[key]
    }
    return F(-math.MaxInt32, 0)
}
